Email sent to author upon "DataPending" status needs to be customisable

luistoptal opened the pull request:
# Pull request for issue: #1618

This is a pull request for the following functionalities:

* If editor sets status to DataPending, they are going to see a modal popup that prompts them for the email sent to the author

## How to test?

* Go to edit any dataset entry
* Set status from, e.g. "submitted" to "DataPending" and save, you are expected to see a modal popup (you might need to follow the fuw workflow):
![image](https://github.com/gigascience/gigadb-website/assets/143437854/a33de1b8-5b21-4ad9-ac37-dab590f26297)
* Validate appearance and copy of modal content
* Navigate modal with keyboard
* Update textarea value to anything you want, optionally including `{{ identifier }}`
* Click on "save and send email"
* Verify that a new email with the expected content appears in `fuw/app/frontend/runtime/mail`

## How have functionalities been implemented?

* Different options to implement this feature were evaluated, e.g. dynamically display additional field under upload status field, display persistent field, ... I opted for the current solution because it does not bloat the form, there's no risk that editor forgets about it or misses it and to me it makes sense that from a UI perspective it's not part of the form because it's not a dataset metadata or attribute (but technically the email textarea is part of the form and submitted as formdata along with the rest).
* The "save" button of the form conditionally opens a modal (instead of submitting the form) if the editor has changed the upload status to "DataPending" (from some other value)
* The modal shows an additional input that alllows the editor to customize the email sent to the author. this field is required which means that the editor cannot submit the form with an empty value
* If the "save and send" button in the modal is clicked, the form is submitted with the additional input value
* the controller intercepts the post request. If the formdata is in the "datapending" case and there is an email message defined, the controller will use this email message, parse the identifier placeholder, and submit as email to the author. If this email message is not defined, the controller defaults to the usual behavior (using a twig template)
* the default value for the email message is copied from the twig template and is hardcoded within the script tag
* IMPORTANT: if in production some other twig template is used, this value needs to be updated to whatever value it has in production. Alternatively I can investigate how to get the text from the twig template and show it in the frontend

## Any issues with implementation?

--

## Any changes to automated tests?

--

## Any changes to documentation?

--

## Any technical debt repayment?

--

## Any improvements to CI/CD pipeline?

--

luistoptal committed:
feat: promtp editor to customize email body sent to author

luistoptal committed:
feat: improve style and a11y of email modal

luistoptal committed:
fix: typo

luistoptal committed:
feat: hardcode return focus to modal trigger btn

luistoptal committed:
chore: update changelog

rija commented:
Hi @luistoptal,
Could you rebase this PR, I'll review it after I'm done reviewing the relatively trivial #1738

luistoptal committed:
chore: Merge branch 'develop' into new-feature/1618-email-sent-to-author-upon-datapending-status-needs-to-be-customisable

rija reviewed (commented):
Hi @luistoptal,

**praise**: functionality works in Firefox
**praise**: code looks fine
**issue**: The "Save and send" button doesn't work in Safari
**issue**: the default template being in the view code is not consistent with our practice of using email templates in Twig format for all outgoing communication. 
**suggestion**: could we not read the Twig template from the javascript code? I think the template path (`files/templates`) is web accessible (which is problematic but we can tighten that to limit access to the domain only later)
**issue** (non-blocking): I don't think it's a good idea to have a custom templating function (`processTemplateString`) either.
I do understand that you could not use the existing `renderNotificationEmailBody` from DatasetUpload.php as that function load the default text from the twig template.

**note**: one solution could be to add additional template loader to chain in the Twig environment. In our case what comes from the form data is a string, so we can use [twig-string-loader](https://packagist.org/packages/shapecode/twig-string-loader), but that's something we can do in the backend as we would want to provide that to all the area that send email from an admin form action, so don't worry about that for now

luistoptal commented:
@rija I will take a look at this on 29th

luistoptal committed:
fix: safari compat to detect clicked submit btn

luistoptal committed:
chore: typo

luistoptal committed:
refactor: fetch template string directly from twig file

luistoptal commented:
> The "Save and send" button doesn't work in Safari

Fixed. Safari didn't detect focused submit button as expected, added support for Safari's different behavior

> could we not read the Twig template from the javascript code

I updated the code to fetch the twig template text from `/files/templates/DataPending.twig`

luistoptal committed:
chore: Merge branch 'develop' into new-feature/1618-email-sent-to-author-upon-datapending-status-needs-to-be-customisable

rija reviewed (commented):
Hi @luistoptal,

**praise**: I can send custom email with Safari when setting status to "DataPending"
**praise**: I can send custom email with Firefox when setting status to "DataPending"
**praise**: I can see that the template is loaded from the Twig file
**issue**: The link "Reset to default email template" doesn't work (on any web browser) and throw a an error in the JS console `Uncaught ReferenceError: defaultDataPendingEmailBody is not defined`
**suggestion**: maybe load the twig template into `defaultDataPendingEmailBody` and use that variable wherever the template content is needed
**issue** (non-blocking): the look and feel of the email editor is different from that of the custom email editor that popup when clicking the dropbox icon in the dataset admins list after setting a dataset to "AssigningFTPbox" status.
**suggestion**: make both editor looks and behave the same, but it's non-blocking because there is already a ticket that's more appropriate for that work probably: #1766? (although that ticket was about CSS weirdness primarily). Up to you where you feel is more appropriate to look into this
**issue** (non-blocking): After setting the status to "DataPending" and an email is sent, if we then set the status to something else (e.g: "DataAvailableForReview"), end then set the status to "DataPending" again, the email won't be sent after clicking "Save and send" and although the status is set to "DataPending" correctly, it's redirected to the dataset list instead of staying on the form.
**suggestion**: it's going to be a problem as there's quite a few back-and-forth interaction between authors and curators and we should be able to set to "data pending" many time. The problem is that the workflow is coded to expected a transition from Submitted and DataPending only (see `setStatusToDataPending` in `DatasetUpload.php`) so there's more implications that need taken care of and it's backend work. So you can ignore that issue in the context of this PR. I will create a ticket (or repurpose an existing one if appropriate) to deal with this.
**praise**: All tests are passing on my local environment
**praise**: All tests are passing on Gitlab CI

luistoptal committed:
fix: set default template on button click

luistoptal commented:
Hi @rija 

> The link "Reset to default email template" doesn't work

Good catch, I missed that when I applied the previous fix

> the look and feel of the email editor is different from that of the custom email editor that popup when clicking the dropbox icon in the dataset admins list after setting a dataset to "AssigningFTPbox" status

There's another PR that changes the look of that dialog with respect to how it looks in this PR (https://github.com/gigascience/gigadb-website/pull/1699), they differ in the input label (one is bold, the other is nromal text). There's also a Tip box but that difference seems suitable as the content within the tip does not describe the input whether the one in this PR does

> then set the status to "DataPending" again, the email won't be sent after clicking "Save and send" and although the status is set to "DataPending" correctly

I see what you mean. I think there's a slight coupling between the front and the back that determines when the email is sent from the backend / prompted to the user in the front

Front: user loads the page with status other than data pending, then sets status to data pending
Back: only if status is switched from submitted to data pending, email is sent

I feel like it's a bit awkward to have this logic duplicated in front and back (and not matching at this point). I thnk it would be nice to have this condition in one place. From what you say it looks like the front has the expected condition so I will not make any change, so I agree this can be left to its own ticket

luistoptal committed:
chore: Merge branch 'develop' into new-feature/1618-email-sent-to-author-upon-datapending-status-needs-to-be-customisable

rija committed:
Merge branch 'develop' into new-feature/1618-email-sent-to-author-upon-datapending-status-needs-to-be-customisable

rija reviewed (approved):
Hi @luistoptal,

Based on the changes and the comments, I'm happy to approve this PR

I believe the backend/frontend dissonance will fix itself when #1616 is implemented
